### PR TITLE
🍰 Send out notifications on CreateComment

### LINKED ISSUE #1017: 🚀 [Feature] Send out notifications on CreateComment
## :rocket: Feature

If a user mentions an other user in a Comment:
- He should get a notification.
  - If the user clicks on the notification where he/she is mentioned he/she should get to the Post page and the comment should be shown.
- If one clicks on the link of the Mention, one should get to the Users page. 

### Additional context

The notification should redirect to the Comment on the related Post page via a query in the URL that uses the Comment ID.

### PR DESCRIPTION
**Pull request by Tirokk**
## 🍰 Pullrequest

Implement notifications in Comments.

### Issues
- fixes #1017

### Todo

- [X] avoid Hashtags in Comments
- [x] extract and save notifications to the database
- [x] implement click on Comment notification and go to the Comment
- [x] implement Frontend tests
- [x] implement Backend tests

### CODE REVIEW

File: webapp/nuxt.config.js [hunk 1]
```diff
@@ -106,17 +121,98 @@ module.exports = {
...
+      }
+
+      // savedPosition is only available for popstate navigations (back button)
+      if (savedPosition) {
+        position = savedPosition
+      }
+
+      return new Promise(resolve => {
+        // wait for the out transition to complete (if necessary)
+        window.$nuxt.$once('triggerScroll', () => {
```
Comment (roschaefer) [inline webapp/nuxt.config.js]: I hope this is SSR safe! (although considering it's the `nuxt.config.js` it should be the case)

File: webapp/pages/post/_id/_slug/index.vue [hunk 2]
```diff
@@ -110,10 +110,12 @@ export default {
     }
   },
   data() {
+    // Wolle: const { commentId = null } = this.$route.query
```
Comment (roschaefer) [inline webapp/pages/post/_id/_slug/index.vue]:
```suggestion
```

File: webapp/pages/post/_id/_slug/index.vue [hunk 3]
```diff
@@ -110,10 +110,12 @@ export default {
...
     return {
       post: null,
       ready: false,
       title: 'loading',
+      // Wolle: commentId,
```
Comment (roschaefer) [inline webapp/pages/post/_id/_slug/index.vue]:
```suggestion
```

File: webapp/nuxt.config.js [hunk 4]
```diff
@@ -57,11 +57,26 @@ module.exports = {
...
+        name: 'description',
+        content: pkg.description,
+      },
+    ],
+    link: [
+      {
+        rel: 'icon',
+        type: 'image/x-icon',
+        href: '/favicon.ico',
+      },
```
Comment (roschaefer) [inline webapp/nuxt.config.js]: Unrelated?

File: webapp/components/notifications/NotificationMenu/index.vue [hunk 5]
```diff
@@ -17,30 +17,80 @@
...
 
-const NOTIFICATIONS = gql(`{
-  currentUser {
-    id
-    notifications(read: false, orderBy: createdAt_desc) {
-      id read createdAt
-      post {
-        id createdAt disabled deleted title contentExcerpt slug
-        author { id slug name disabled deleted }
+const NOTIFICATIONS = gql`
```
Comment (roschaefer) [inline webapp/components/notifications/NotificationMenu/index.vue]: Move that into `~/graphql`

File: webapp/components/notifications/Notification/Notification.vue [hunk 6]
```diff
@@ -42,11 +51,26 @@ export default {
...
+      return this.notification.comment || {}
+    },
+    postParams() {
+      return {
+        id: this.post.id || this.comment.post.id,
+        slug: this.post.slug || this.comment.post.slug,
+      }
+    },
+    hashParam() {
+      return this.post.id ? {} : { hash: `#commentId-${this.comment.id}` }
```
Comment (roschaefer) [inline webapp/components/notifications/Notification/Notification.vue]: I was thinking about this lately: How about a generic behaviour for fragment identifiers? Can't we simply scroll to *any* html element which has a `scroll-id="blabla"`? We could use it for comments but maybe for other use-cases as well?

File: webapp/components/notifications/Notification/Notification.vue [hunk 7]
```diff
@@ -42,11 +51,26 @@ export default {
   },
   computed: {
     excerpt() {
-      return this.$filters.removeLinks(this.post.contentExcerpt)
+      const excerpt = this.post.id ? this.post.contentExcerpt : this.comment.contentExcerpt
+      return (
+        (!this.post.id ? '<b>Comment: </b>' : '') + excerpt.replace(/<(?:.|\n)*?>/gm, '').trim()
```
Comment (roschaefer) [inline webapp/components/notifications/Notification/Notification.vue]: Can we implement this behaviour in the template please?
```vue
<div v-if="resourceType == 'Post'">
  {{ contentExcerpt | removeHtml }}
</div>
<b v-else>
  {{ contentExcerpt | removeHtml }}
</b>
```

File: webapp/components/notifications/Notification/Notification.vue [hunk 8]
```diff
@@ -42,11 +51,26 @@ export default {
...
+      )
     },
     post() {
       return this.notification.post || {}
     },
+    comment() {
+      return this.notification.comment || {}
+    },
+    postParams() {
```
Comment (roschaefer) [inline webapp/components/notifications/Notification/Notification.vue]: Well, it's not `postParams` anymore right? Just `params` maybe?

File: webapp/components/notifications/Notification/Notification.vue [hunk 9]
```diff
@@ -2,20 +2,29 @@
...
-        <hc-user :user="post.author" :date-time="post.createdAt" :trunc="35" />
+        <hc-user
+          :user="post.author || comment.author"
+          :date-time="post.createdAt || comment.createdAt"
+          :trunc="35"
+        />
       </ds-space>
       <ds-text color="soft">
-        {{ $t('notifications.menu.mentioned') }}
+        {{ $t('notifications.menu.mentioned', { resource: post.id ? 'post' : 'comment' }) }}
```
Comment (roschaefer) [inline webapp/components/notifications/Notification/Notification.vue]:
```suggestion
        {{ $t('notifications.menu.mentioned', { resourceType }) }}
```

File: webapp/components/notifications/Notification/Notification.vue [hunk 10]
```diff
@@ -2,20 +2,29 @@
   <ds-space :class="{ notification: true, read: notification.read }" margin-bottom="x-small">
     <no-ssr>
       <ds-space margin-bottom="x-small">
-        <hc-user :user="post.author" :date-time="post.createdAt" :trunc="35" />
+        <hc-user
+          :user="post.author || comment.author"
+          :date-time="post.createdAt || comment.createdAt"
+          :trunc="35"
+        />
```
Comment (roschaefer) [inline webapp/components/notifications/Notification/Notification.vue]: Maybe:
```vue
        <hc-user v-if="resourceType == 'Post'"
          :user="post.author"
          :date-time="post.createdAt"
          :trunc="35"
        />
        <hc-user v-else
          :user="comment.author"
          :date-time="comment.createdAt"
          :trunc="35"
        />
```

Comment (roschaefer) [reply webapp/components/notifications/Notification/Notification.vue]: Alternatively you could use a computed property:
```js
computed: {
  mentioner() {
    return this.post.id ? this.post.author : this.comment.author
  }
}
```

File: backend/src/middleware/handleHtmlContent/handleContentData.js [hunk 11]
```diff
@@ -1,39 +1,46 @@
...
   const createdAt = new Date().toISOString()
   const cypher = `
-    MATCH(p:Post {id: $postId})<-[:WROTE]-(author:User)
-    MATCH(u:User)
+    MATCH (source)
+    WHERE source.id = $id AND $label IN LABELS(source)
+    MATCH (source)<-[:WROTE]-(author: User)
+    MATCH (u: User)
     WHERE u.id in $idsOfMentionedUsers
     AND NOT (u)<-[:BLOCKED]-(author)
```
Comment (roschaefer) [inline backend/src/middleware/handleHtmlContent/handleContentData.js]: The `author` must be different here. Consider the following situation:

1. (personA)-[:BLOCKED]->(personB)
2. (personA)-[:WROTE]->(post)
3. (personC)-[:WROTE]->(comment)-[:COMMENTED]->(post)

Now what happens if `personC` would mention `personB` in post authored by `personA`? As `personB` you would get a notification linked to a post and following it you get a 404. **That's** the situation we want to avoid.

It is no problem right now if the author of a comment mentions a blocked user. The blocked user could still follow the link and not get a 404. We would change this behaviour as soon as your blocked users cannot see your comments anymore. But I don't think that we will implement this behaviour, because you would have totally different conversation flows per user account.

Comment (roschaefer) [reply backend/src/middleware/handleHtmlContent/handleContentData.js]: Could you write a backend test for this? Have a look into [this example here](https://github.com/Human-Connection/Human-Connection/blob/master/backend/src/middleware/handleHtmlContent/handleContentData.spec.js#L170).

File: backend/src/middleware/handleHtmlContent/handleContentData.js [hunk 12]
```diff
@@ -56,15 +63,29 @@ const handleContentData = async (resolve, root, args, context, resolveInfo) => {
...
+}
+
 export default {
   Mutation: {
-    CreatePost: handleContentData,
-    UpdatePost: handleContentData,
+    CreatePost: handleContentDataOfPost,
+    UpdatePost: handleContentDataOfPost,
+    CreateComment: handleContentDataOfComment,
+    // UpdateComment: handleContentDataOfComment,
```
[the remark this answers is not in the document]
Comment (mattwr18) [reply backend/src/middleware/handleHtmlContent/handleContentData.js]: this is fixed in a subsequent `PR`, although I guess it could be added here as well @roschaefer

File: backend/src/middleware/handleHtmlContent/handleContentData.js [hunk 13]
```diff
@@ -56,15 +63,29 @@ const handleContentData = async (resolve, root, args, context, resolveInfo) => {
...
 
-  await notify(post.id, idsOfMentionedUsers, context)
+  await notifyMentions('Post', post.id, idsOfMentionedUsers, context)
   await updateHashtagsOfPost(post.id, hashtags, context)
 
   return post
 }
 
+const handleContentDataOfComment = async (resolve, root, args, context, resolveInfo) => {
+  // extract user ids before xss-middleware removes classes via the following "resolve" call
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply backend/src/middleware/handleHtmlContent/handleContentData.js]: I removed this comment and similar ones …

File: backend/src/middleware/handleHtmlContent/handleContentData.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (Tirokk) [reply backend/src/middleware/handleHtmlContent/handleContentData.js]: I forgot to remove this comment signs …
But did now. 🤗
You can smile again 😆

File: backend/src/seed/seed-db.js [hunk 14]
```diff
@@ -44,42 +44,49 @@ import Factory from './factories'
...
         email: 'user@example.org',
       }),
       f.create('User', {
         id: 'u4',
-        name: 'Tick',
+        name: 'Huey (Tick)',
+        slug: 'huey-tick',
         role: 'user',
-        email: 'tick@example.org',
+        email: 'huey@example.org',
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply backend/src/seed/seed-db.js]: Good point!
Done 😍

File: webapp/pages/post/_id/_slug/index.vue (hunk 2 above)
Comment (Tirokk) [reply webapp/pages/post/_id/_slug/index.vue]: Ups.
I forgot to clean up. Sorry! 😚
Done in my work tree …

File: webapp/pages/post/_id/_slug/index.vue (hunk 3 above)
Comment (Tirokk) [reply webapp/pages/post/_id/_slug/index.vue]: Same same …

File: webapp/components/Editor/Editor.vue [hunk 15]
```diff
@@ -204,12 +204,173 @@ export default {
...
+    users: { type: Array, default: () => null }, // If 'null', than the Mention extention is not assigned.
+    hashtags: { type: Array, default: () => null }, // If 'null', than the Hashtag extention is not assigned.
     value: { type: String, default: '' },
     doc: { type: Object, default: () => {} },
   },
   data() {
+    // Set array of optional extensions by analysing the props.
+    let optionalExtensions = []
+    // Don't change the following line. The functionallity is in danger!
+    if (this.users) {
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply webapp/components/Editor/Editor.vue]: Yes, for sure:
In the comments hashtags are not used.

In the future:
If you write a description of yourself in your user profile hashtags and mentions should not be used.

File: webapp/nuxt.config.js (hunk 1 above)
Comment (Tirokk) [reply webapp/nuxt.config.js]: One would hope so … 😉

File: webapp/nuxt.config.js (hunk 4 above)
Comment (Tirokk) [reply webapp/nuxt.config.js]: Linting changed!

File: webapp/components/notifications/NotificationMenu/index.vue (hunk 5 above)
Comment (Tirokk) [reply webapp/components/notifications/NotificationMenu/index.vue]: Done +++ …

File: webapp/components/notifications/Notification/Notification.vue (hunk 6 above)
Comment (Tirokk) [reply webapp/components/notifications/Notification/Notification.vue]: You maybe mix up queries with hashtags here? @roschaefer 

A hashtag is always about scrolling to an anchor. Isn’t it?
The naming `commentId` is for the user readability. That they can get a cloud about what the hashtag is, even in case the scrolling does not work for any reason.

File: webapp/components/notifications/Notification/Notification.vue (hunk 7 above)
Comment (Tirokk) [reply webapp/components/notifications/Notification/Notification.vue]: I don’t really understand this:

- We add `<b>Comment: </b>` there.
- Therefore the naming `excerpt` should be changed to `contentInformation`. Which I have already done now!
- In your example a `|` operator is used and I can not find its functionality.
- The if and else case has the same code … ??? But you mean `comment.contentExcerpt` for the second one I suppose …

Comment (roschaefer) [reply webapp/components/notifications/Notification/Notification.vue]: Let me explain:
* the `v-else` branch is using `<b>` instead of `<div>`
* the `removeHtml` is a filter that needs to be implemented -it's easy :wink:

Comment (Tirokk) [reply webapp/components/notifications/Notification/Notification.vue]: So I found out and succeed. 👍🏼

File: webapp/components/notifications/Notification/Notification.vue (hunk 8 above)
Comment (Tirokk) [reply webapp/components/notifications/Notification/Notification.vue]: Yes, you are right 🤗
Renamed this …

File: webapp/components/notifications/Notification/Notification.vue (hunk 9 above)
Comment (Tirokk) [reply webapp/components/notifications/Notification/Notification.vue]: I did and it gets refactored even more in the next PR 🍰 Send out notifications on CreateComment #1018

File: webapp/components/notifications/Notification/Notification.vue (hunk 10 above)
Comment (Tirokk) [reply webapp/components/notifications/Notification/Notification.vue]: If done the first approach …

File: backend/src/middleware/handleHtmlContent/handleContentData.js (hunk 11 above)
Comment (Tirokk) [reply backend/src/middleware/handleHtmlContent/handleContentData.js]: Done 😆

**Comment by roschaefer** [comment]
Oh no, conflicting files!

**Comment by mattwr18** [comment]
I'll work through them and push it up @roschaefer @Tirokk

**Comment by mattwr18** [comment]
> OK! I see one improvement but I think it's for another day: We would just remove the code complexity by refactoring out the if-clause. My suggestion: As soon as we do something with hashtags in comments (why not tagging comments?) we can refactor the cypher statements and `notify` and `tag` both comments and posts in the same way.
> 
> that'll do

I agree @roschaefer, we talked about this, and even tried to get it working with one `cypher` query, but also we could refactor it later, unless it's obvious to someone how we can do it now.

File: webapp/components/Editor/Editor.vue [hunk 16]
```diff
@@ -200,12 +200,173 @@ export default {
...
+      optionalExtensions.push(
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply webapp/components/Editor/Editor.vue]: I tried under @appinteractive advice to get this in one class, and spend many hours on it, but my JS skills seemed to weren’t enough to get it working … @alina-beck

File: webapp/components/Comment.vue [hunk 17]
```diff
@@ -10,7 +10,7 @@
     </ds-card>
   </div>
   <div v-else :class="{ comment: true, 'disabled-content': comment.deleted || comment.disabled }">
-    <ds-card>
+    <ds-card :id="`commentId-${comment.id}`">
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply webapp/components/Comment.vue]: @mattwr18 and I did it for the user experience. That a user would have a chance for a clue about what this is, even in case the scroll might not working. @alina-beck

File: cypress/integration/common/steps.js [hunk 18]
```diff
@@ -363,95 +363,106 @@ Then("there are no notifications in the top menu", () => {
...
+  cy.openPage("/profile/annoying-user/spammy-spammer");
+});
+
+When("I visit the profile page of the annoying user", name => {
+  cy.openPage("/profile/annoying-user");
+});
+
+When("I ", name => {
+  cy.openPage("/profile/annoying-user");
+});
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply cypress/integration/common/steps.js]: We didn’t do and haven't seen this, but you are absolut right, in my eyes … @alina-beck

File: backend/src/schema/resolvers/notifications.spec.js [hunk 19]
```diff
@@ -52,80 +59,220 @@ describe('currentUser { notifications }', () => {
...
+          factory.create('Notification', {
+            id: 'comment-mention-not-for-you',
+          }),
+          factory.create('Notification', {
+            id: 'comment-mention-already-seen',
+            read: true,
+          }),
+          factory.create('Notification', {
+            id: 'comment-mention-unseen',
+          }),
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply backend/src/schema/resolvers/notifications.spec.js]: Inspired by the earlier ones by @roschaefer … @alina-beck

**Comment by Tirokk** [comment]
@alina-beck The scroll behaviour depends: Often it does work and sometimes does not. That is related to the fact as an example if you are already on the post page or not.
@mattwr18 and I have spend hours to days to get it kind of working.

I would love if you are able to get it working all the time !!! 🤗😍

But later, I suggest, as an error correction. Just to get this merged and start testing as soon as possible.

File: webapp/components/notifications/Notification/Notification.vue [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+          hover
+          space="x-small"
+          class="notifications-card"
+        >
+          <ds-space margin-bottom="x-small" />
+          <div v-if="resourceType == 'Post'">{{ post.contentExcerpt | removeHtml }}</div>
+          <div v-else>
+            <b>
+              Comment: &nbsp;
+            </b>
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply webapp/components/notifications/Notification/Notification.vue]: That’s okay with me, in case @mattwr18 likes to do it, because he will be responsible for today … @alina-beck

File: webapp/components/notifications/Notification/Notification.vue [hunk 21]
```diff
@@ -0,0 +1,89 @@
+<template>
+  <ds-space :class="{ notification: true, read: notification.read }" margin-bottom="x-small">
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply webapp/components/notifications/Notification/Notification.vue]: That’s okay with me, in case @mattwr18 likes to do it, because he will be responsible for today … @alina-beck

File: webapp/components/Editor/Editor.vue (hunk 16 above)
[the remark this answers is not in the document]
Comment (Tirokk) [reply webapp/components/Editor/Editor.vue]: But we would need still the possibility of this options … @alina-beck 

I would love if you show me how to get this in one class !!! 🤗😍

But later, I suggest, as a refactoring. Just to get this merged and start testing as soon as possible.

File: cypress/integration/common/steps.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (mattwr18) [reply cypress/integration/common/steps.js]: sorry @alina-beck these are just linting changes, and make it confusing for the `PR` reviewer

File: webapp/components/Editor/Editor.vue (hunk 16 above)
[the remark this answers is not in the document]
Comment (mattwr18) [reply webapp/components/Editor/Editor.vue]: I agree, we can refactor later, if that's ok

File: cypress/integration/common/steps.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (alina-beck) [reply cypress/integration/common/steps.js]: Oh, you're right – sorry... I didn't check the file name and was already wondering what this PR had to do with annoying users :wink:
